doc/rst/language/spec/data-parallelism.rst

```diff
@@ -751,6 +751,45 @@ side array expressions alias the left-hand side expression.
    are assigned to ``A`` may be read to compute the sum depending on the
    number of tasks used to implement the data parallel statement.
 
+.. _Indirect_Whole_Array_Indexing:
+
+Indirect Whole Array Indexing
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Arrays can be indexed indirectly using an array of indices.
+Given an array ``A`` and an array of indices ``B`` in the domain of ``A``,
+the function call
+
+.. code-block:: chapel
+
+   f(A[B]);
+
+is equivalent to
+
+.. code-block:: chapel
+
+   [b in B] f(A[b]);
```

Comment:
A couple of pieces of feedback:

* It seems like the presence of `f()` is not important in the above example (while also raising unfortunate questions like "What argument type(s) does `f()` accept?").  Specifically, it seems orthogonal to the main point, which I think of as being that the expression `...A[B]...` is equivalent to `...[b in B] A[b]...`.  In particular, I think the more precise way of writing the `f(A[B])` equivalence would be [`f([b in B] A[b])`](https://ato.pxeger.com/run?1=m70sOSOxIDVnwYKlpSVpuhY3jxQU5ScrpOXna1RYKRSlJuZoKlRzKQBZJaVFeQoV1ly1XFxliUUKjgq2CtGGekY6CsZ6JjoKpnpmsdZgCSewhI6CgY4CUNIQKAoWdgYKg0x1jHaK1bTmKi_KLEnNydNw1lFQslJQ0lFw1iupLEi1Ki4pysxL14RqcoFqik5SyMxTcIpVcIxOQtbtAtPtgqob4heol2BeAwA) rather than the rewrite shown here (though assuming `f()` accepts `A.eltType`, a subsequent promotion rewrite would end up with something like the text in the PR).

* To me, this feels like something that doesn't really require its own special spec section because it feels more like "just another application of the promotion concept" (albeit a potentially non-obvious one).  So, while I'd preserve this discussion and example, I'd make it more of a sidebar/discussion that illustrates a potentially non-obvious use of promotion than a section of its own, which suggests it's a new/different/special rule.

  Along those same lines, the current wording mentions it being a promotion only as an adjective along the way rather than calling out that promotion is _why_ this works and behaves in this way, which seems important (e.g., imagine leading with something like "Array indexing operations can also be promoted.  For example, an array of indices can be used to index into another array, resulting in the following equivalence...").

  Putting these sentiments another way, anything that can promote to an array's index type should be able to be used to promote array indexing, so focusing on arrays being indexed by arrays seems overly specific as a spec rule vs. as an illustration of promoted indexing.  However, this distinction I'm making isn't deeply relevant today because ranges and domains are the two other main (only?) things that can promote, and they already tend to have special behaviors due to their role in array slicing (which, by design, behaves similarly to promoted indexing in a closed/compact form).  That said, once we support promotion by other data structures (e.g., lists, user-defined data structures, etc.), those should be able to be used to promote as well, and without requiring their own special sections saying so.  Even today, I think it ought to be possible to promote a rectangular domain `A` using an associative domain of `index(A.domain)` (say), but I think that it doesn't work as I'd expect because we incorrectly fall into the normal array slicing cases and then fail rather than promoting as I'd expected...(?)